Initial Implementation of Tasks in CMS Configurations

wddgit opened the pull request:
This includes the initial implementation of the Task feature in the Core CMS configuration code. It allows creating Tasks, adding modules to them, and associating them to sequences, paths, end paths, and the schedule. In many ways Tasks behave like Sequences except that the order modules appear on Tasks does not matter. New unit tests were added to test the Core code changes.

This PR also includes changes to migrate a lot of configuration code to use Tasks. Large parts of PAT configuration code, all PAT tools, and some parts of RECO code were migrated. In particular, code that was used by runTheMatrix.py tests, IB unit tests, and other test configurations and that was already using unscheduled execution mode was migrated. Although large numbers of configuration files were migrated, an even larger number of configuration files were not migrated yet. There remains much migration work to be done. But the unmigrated configurations are not broken by this PR and should continue to run successfully. At a future date, the other code can be incrementally migrated as resources become available to do the work.

This PR modifies the convertToUnscheduled function to work properly with the new Task feature. EDProducers and EDFilters removed from paths and endpaths are automatically placed on a Task.

There are also changes to ConfigBuilder.py.

This is the third PR submitted for this set of changes. The first 2 were closed to make significant revisions. The first one was PR #16538 and there are many comments from Chris in that PR which have all been addressed in separate commits after the first 3 commits. The second one was PR #16925, which contains Slava's comments and a lot of discussion that followed. (Also PR#17430 which was the auto port to 9_1_X when it opened)

wddgit committed:
Initial implementation of Task in the python configuration language

wddgit committed:
Changes to make Core unit tests work with the new Task feature

wddgit committed:
Make the runTheMatrix tests and unit tests outside the Core pass with Tasks

wddgit committed:
Add parentheses to improve code readability

wddgit committed:
Fix bugs where dumpPython was writing function calls with more than 255 arguments

wddgit committed:
Change function name to indicate it is private

wddgit committed:
Add more info to exception message

wddgit committed:
Move testTask_cff from python to python/test directory

wddgit committed:
Use OrderedSet instead of a list and a set in Task, _ModuleSequenceType, and Schedule. There should be no change in behavior.

wddgit committed:
Delete some test code unintentionally committed moments earlier

wddgit committed:
Use one DependenceOrder function for Tasks and Sequences instead of two

wddgit committed:
Rename and fix allowUnscheduled unit tests

wddgit committed:
Associate PAT task directly with Schedule instead of new PAT EndPath

wddgit committed:
Use new function to which calls both setattr and Task::add

wddgit committed:
Migrate PatAlgos tests to be consistent with last couple commits

wddgit committed:
Bug fix so last 3 commits compile

wddgit committed:
Delete extraneous lines mistakenly added that do nothing

wddgit committed:
Migrate new test recently added to CMSSW to use a Task

wddgit committed:
After rebasing on 9_0_X on 1/10/2017 there was new code that needed to be migrated to Tasks

wddgit committed:
Modify so that sequences are initialized by associating to a Task

The purpose is to eliminate the maintenance burden of maintaining
nearly duplicate lists of items in Sequences and Tasks. Note that
before this commit the modules that would be setup to run unscheduled
was intentionally identical to what it was before the Pull Request,
but after the commit this is no longer true. In places where
the existing list of unscheduled modules and the existing sequences
differed, decisions had to be made to reconcile them. (It was
my original intent for this to happen, but before this commit
I intended it as a second step in the migration after the PR)

Also rebased the entire PR again when I pushed this commit.

wddgit committed:
Some necessary additions or tests fail

wddgit committed:
Fix problems with module order that showed up in addOnTests

wddgit committed:
Fix problem revealed by a unit test.

wddgit committed:
Fix a module ordering problem in runTheMatrix 25.0 step 3

wddgit committed:
Migrate more code to use Tasks

wddgit committed:
Fix PatAlgos helpers.py unit test so it runs and passes

wddgit committed:
MassReplace functions also modify tasks associated to Schedule

wddgit committed:
Fix test configurations not run by unit tests

wddgit committed:
Fix conflicts and migrate recently merged code to Tasks

wddgit committed:
Migrate a new test configuration to use Tasks

wddgit committed:
Migrate new PAT tools code to Tasks

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for master.

It involves the following packages:

CommonTools/ParticleFlow
Configuration/Applications
Configuration/StandardSequences
DQMOffline/JetMET
DataFormats/FWLite
EgammaAnalysis/ElectronTools
FWCore/Concurrency
FWCore/Framework
FWCore/Integration
FWCore/ParameterSet
FWCore/PythonParameterSet
HLTrigger/Configuration
HeavyFlavorAnalysis/RecoDecay
HeavyFlavorAnalysis/Skimming
HeavyFlavorAnalysis/SpecificDecay
IOPool/Common
IOPool/Output
JetMETCorrections/Configuration
JetMETCorrections/Type1MET
PhysicsTools/Heppy
PhysicsTools/JetMCAlgos
PhysicsTools/PatAlgos
PhysicsTools/PatExamples
PhysicsTools/PatUtils
PhysicsTools/SelectorUtils
RecoBTag/CTagging
RecoBTag/Combined
RecoBTag/Configuration
RecoBTag/ImpactParameter
RecoBTag/SecondaryVertex
RecoBTag/SoftLepton
RecoEgamma/ElectronIdentification
RecoEgamma/PhotonIdentification
RecoJets/JetProducers
RecoMET/METPUSubtraction
RecoMuon/MuonIdentification
RecoMuon/MuonIsolation
RecoTauTag/Configuration
RecoTauTag/RecoTau
RecoVertex/AdaptiveVertexFinder
TopQuarkAnalysis/Configuration
TopQuarkAnalysis/Examples
TopQuarkAnalysis/TopEventProducers
TopQuarkAnalysis/TopEventSelection
TopQuarkAnalysis/TopHitFit
TopQuarkAnalysis/TopJetCombination
TopQuarkAnalysis/TopKinFitter
TopQuarkAnalysis/TopPairBSM
TopQuarkAnalysis/TopSkimming
TopQuarkAnalysis/TopTools
Validation/RecoB
Validation/RecoJets
Validation/RecoMET


@perrotta, @smuzaffar, @Dr15Jones, @silviodonato, @fwyzard, @franzoni, @dmitrijus, @Martin-Grunewald, @cmsbuild, @slava77, @vanbesien, @monttj, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @TaiSakuma, @gouskos, @yslai, @felicepantaleo, @abbiendi, @rappoccio, @Martin-Grunewald, @ahinzmann, @geoff-smith, @mmarionncern, @kreczko, @JyothsnaKomaragiri, @makortel, @acaudron, @jhgoh, @lgray, @jdolen, @HuguesBrun, @ferencek, @trocino, @rociovilar, @Sam-Harper, @wmtan, @GiacomoSguazzoni, @rafaellopesdesa, @rovere, @VinInn, @bellan, @nhanvtran, @gkasieczka, @jalimena, @schoef, @dgulhan, @echapon, @battibass, @calderona, @mverzett, @cbernet, @gpetruc, @imarches, @mariadalfonso, @pvmulder, @bachtis this is something you requested to watch as well.
@muzaffar, @davidlange6, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

wddgit commented:
This is the new PR. It was rebased to the same commit as CMSSW_9_1_X_2017-03-05-2300. And some recently merged PAT Tools code was migrated to use tasks. That is the only difference compared with the previous PR.

Please close the auto-forwarded PR ( #17726  ).  Note that there have been no negative commits to this for weeks and as far as I know all issues have been resolved.

Dr15Jones commented:
please test

Dr15Jones commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 81d1e152a42ea72c05953a758fbd806e910a0ef1

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17801/18178/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS
---> test testPythonParameterSet had ERRORS

cmsbuild commented:
Comparison job queued.

Dr15Jones commented:
Everyone, please review this pull request. It has required constant updates to keep up with other changes. The sooner it gets integrated the sooner developers will begin to learn about Tasks. It would be best to get it in early in 9_1.

dmitrijus commented:
+1

wddgit commented:
The unit test failures are unrelated to this PR.

fwyzard commented:
the only HLT-related difference seems to be in the definition of `MassReplaceInputTag` and `MassReplaceParameter`

fwyzard commented:
+1

Dr15Jones commented:
@davidlange6 how can we move forward with this pull request?

Dr15Jones commented:
@slava77 @perrotta are there concerns from reconstruction?

slava77 commented:
On 3/13/17 6:19 AM, Chris Jones wrote:
> @slava77 <https://github.com/slava77> @perrotta
> <https://github.com/perrotta> are there concerns from reconstruction?

Can we get to this until after pre1 is built?

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17801#issuecomment-286104235>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbv7CFfUCxAeKoY9Ug1dGPz3kiQSIks5rlUJEgaJpZM4MUjDj>.
>

Dr15Jones commented:
I had hoped this would make pre1, but I understand if you are overloaded.

davidlange6 commented:
could we agree on a timescale (since the date of pre1 is in any case uncertain..)

> On Mar 13, 2017, at 2:51 PM, Chris Jones <notifications@github.com> wrote:
> 
> I had hoped this would make pre1, but I understand if you are overloaded.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
On 3/13/17 6:51 AM, Chris Jones wrote:
> I had hoped this would make pre1, but I understand if you are overloaded.
> 
yes, there is some backlog of PRs aiming for pre1.
If the tasks PR goes in, it will disrupt  probably a half of the pending
 91X PRs.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17801#issuecomment-286112356>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbo_7AI17qQ2i4ku1rnRvkLF7_mxxks5rlUnmgaJpZM4MUjDj>.
>

wddgit commented:
I checked today and there are no conflicts that I need to resolve for us to merge this PR. I skimmed all the code changes on the master branch up to commit bf3678e53599d (at the moment the head of the master branch) and saw no logical conflicts or new tests that require migrating either. No need to fix and rebase this PR again today.

davidlange6 commented:
hi @wddgit @slava77 - sorry we didn't manage to discuss this in the ORP - I propose that after we get the RECO changes in for pre1 (tomorrow..) - we'll merge this and make pre1. Given the schedule it should not miss pre1 imho

slava77 commented:
@cmsbuild please test
to check that nothing bad has accumulated since the last test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on JetMETCorrections/Configuration/python/JetCorrectorsAllAlgos_cff.py:
```diff
@@ -520,666 +560,909 @@
+
+jetCorrectorsAllAlgosTask = cms.Task(
+    jetCorrectorsTask,
+    ak7L1JPTOffsetCorrectorTask,
```
the sorting order is somewhat arguable.

I'd group by the first part, preceding LX\*CorrectorTask
and maybe even leave a commented out ```xyzLX\*CorrectorTask``` to be used as a template for other jet algorithms.
We more frequently add an jet algorithm than a new corrector

Also, some separator between subsets (an empty line or ```#LX\*CorrectorTask follow``` or similar) would help with readability.

@schoef @aperloff

slava77 replied:
Also, somebody should eventually refactor this file to be less verbose and multiplex the used elements with some python methods

aperloff replied:
It has been on my to-do list for quite some time, though as no one as complained to me it hasn't floated to the top. It seems to me like this initial PR should go through and then we can revisit the refactoring. It will take me some time to learn about Tasks.

slava77 commented on JetMETCorrections/Type1MET/python/correctionTermsCaloMet_cff.py:
```diff
@@ -1,13 +1,7 @@
-                                                              ak4CaloL2L3ResidualCorrectorChain, \
-                                                              ak4CaloResidualCorrector, \
-                                                              ak4CaloL2L3ResidualCorrector
+from JetMETCorrections.Configuration.JetCorrectors_cff import *
```
this will apparently import about 40 different chains instead of just 2.
Will they get purged eventually if not used or will we have all the extra objects floating in the python global space?

slava77 commented on PhysicsTools/PatAlgos/python/slimming/metFilterPaths_cff.py:
```diff
@@ -55,3 +55,27 @@ def miniAOD_customizeMETFiltersFastSim(process):
 
+metFilterPathsTask = cms.Task(
+    HBHENoiseFilterResultProducer,
+    HBHENoiseFilter,
```
can this task be generated directly from ```allMetFilterPaths```?
tasks corresponding to individual paths defined above can be added and then automatically placed into the large metFilterPathsTask

I'm trying to reduce the number of distantly connected places where edits should be   made to get a new filter

wddgit replied:
Are the filters in metFilterPaths_cff.py ever meant to run unscheduled or are they always supposed to be actual filters on a Path? At least the one filter I checked actually produced a bool so it is possible they are intended to run unscheduled in some use cases, but if they are not then I could just delete this Task entirely.

slava77 replied:
@mmarionncern @zdemirag 
please clarify on this.
Thank you.

wddgit replied:
@mmarionncern @zdemirag I could make this question a little more specific. When metFilterPaths_cff is loaded do we want these filters to be setup by default to run unscheduled? If they are on an explicit Path there is no reason to do this. If we do not want them to run at all, there is no reason to do this. The only thing that loads this file is PhysicsTools/PatAlgos/python/slimming/slimming_cff.py and that file is most often loaded by ConfigBuilder when PAT is included. There are also a small number of test configuration files that explicitly load slimming_cff. Even if we delete the Task from metFiltersPaths_cff, one could still write a test configuration that defines such a Task for some special purpose. The question is do we want these filters to be setup by default for unscheduled whenever PAT is included by ConfigBuilder, and this setup implies the filter modules would be constructed by the C++ part of the Framework.

wddgit commented:
git says no conflicts and I visually skimmed through recent changes up to the head of master (commit 5e507f4) and I cannot see anything that should logically conflict nor any new tests requiring migration. I'll start looking at Slava's comments now.

slava77 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -191,6 +201,10 @@ def miniAOD_customizeCommon(process):
     process.load("RecoJets.JetProducers.PileupJetID_cfi")
+    task.add(process.pileupJetId)
+    task.add(process.pileupJetIdCalculator)
+    task.add(process.pileupJetIdEvaluator)
```
how was this task.add set of lines derived?
Can it come directly from PileupJetID_cfi in some way?
This file doesn't have to know what's in the imported file to reduce maintenance burden

wddgit replied:
How was it derived? I looked inside the cfi file at what the load attaches to the process knowing that those would be the things setup to run unscheduled. Then I made sure the same modules were attached to a task so the list of modules configured to run unscheduled remained the same. (And I ran tests that checked for mistakes)

In this case this is a cfi which originally had no Sequence in it. Do we want Tasks defined in cfi files? I am still thinking about this and we can discuss the rest your question more.

slava77 commented on PhysicsTools/PatAlgos/python/slimming/slimming_cff.py:
```diff
@@ -18,3 +18,25 @@
+slimmingTask = cms.Task(
+    packedPFCandidatesTask,
+    lostTracks,
+    offlineSlimmedPrimaryVertices,
```
this is not easily maintainable.
individual module additions better be replaced by xyzTask following the pattern of loaded files.

Maybe
```
slimmingTask = cms.Task()

 from PhysicsTools.PatAlgos.slimming.packedPFCandidates_cff import *
slimmingTask += packedPFCandidatesTask
```
and so on

slava77 commented on PhysicsTools/PatAlgos/python/tools/jetTools.py:
```diff
@@ -274,99 +349,193 @@ def setupBTagging(process, jetSource, pfCandidates, explicitJTA, pvSource, svSou
-                setattr(process, btagPrefix+btagInfo+labelName+postfix, btag.pfDeepCMVATagInfos.clone(pfDeepCSVTagInfos = cms.InputTag(btagPrefix+'pfDeepCSVTagInfos'+labelName+postfix), ipInfoSrc = cms.InputTag(btagPrefix+"pfImpactParameterTagInfos"+labelName+postfix), muInfoSrc = cms.InputTag(btagPrefix+"softPFMuonsTagInfos"+labelName+postfix), elInfoSrc = cms.InputTag(btagPrefix+"softPFElectron ...
+                addToProcessAndTask(btagPrefix+btagInfo+labelName+postfix,
+                                    btag.pfDeepCMVATagInfos.clone(
+                                        pfDeepCSVTagInfos = cms.InputTag(btagPrefix+'pfDeepCSVTagInfos'+labelName+postfix),
```
thanks for the line-breaks

slava77 commented on PhysicsTools/PatAlgos/python/triggerLayer1/triggerMatcherExamples_cfi.py:
```diff
@@ -180,3 +180,19 @@
 )
+
+triggerMatcherExamplesTask = cms.Task(
+    somePatMuonTriggerMatchHLTMu17,
```
can this be refactored somewhat, e.g. triggerMatcherExamplesTask extended by one module right after each one is added?
Note that with ``` _exampleTriggerMatchers ``` we get the third place in this file to edit if some changes are needed

slava77 commented on PhysicsTools/PatAlgos/test/miniAOD/patTuple_mini_data.py:
```diff
@@ -26,7 +27,12 @@
+patAlgosToolsTask.add(process.slimmingTask)
+
 process.load("RecoVertex.AdaptiveVertexFinder.inclusiveVertexing_cff")
+patAlgosToolsTask.add(process.inclusiveVertexingTask)
```
one load, one task add could be a more clean implementation pattern.
Better yet, if this can be done in one line with a "smarter" method .. although this will probably not always work

slava77 commented on RecoBTag/SoftLepton/python/softLepton_cff.py:
```diff
@@ -8,3 +8,32 @@
+
+softLeptonTask = cms.Task(
+    softPFElectronsTagInfos,
+    softPFMuonsTagInfos,
```
how do the 6 imports map to the 26 modules entered in this task?
Can this be made in a more maintainable way?

slava77 commented on RecoTauTag/Configuration/python/boostedHPSPFTaus_cfi.py:
```diff
@@ -1,13 +1,21 @@
+    patAlgosToolsTask.add(process.pfPileUpForBoostedTaus)
+    patAlgosToolsTask.add(process.pfNoPileUpForBoostedTaus)
+    patAlgosToolsTask.add(process.ca8PFJetsCHSprunedForBoostedTaus)
+    patAlgosToolsTask.add(process.boostedTauSeeds)
```
having a boostedHPSPFTausTask in boostedHPSPFTaus_cff would make it simpler

... there was some lack of attention in review in the past that let this  xyz_cfi  depend on xyz_cff

cmsbuild commented:
-1

Tested at: 81d1e152a42ea72c05953a758fbd806e910a0ef1

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17801/18502/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS
---> test testPythonParameterSet had ERRORS

cmsbuild commented:
Comparison job queued.

slava77 commented:
On 3/16/17 6:03 PM, cmsbuild wrote:
> *Failed tests: UnitTests*
> 
>   * *Unit Tests*:
> 
> I found errors in the following unit tests:
> 
> ---> test testRecoMETMETProducers had ERRORS
> ---> test testPythonParameterSet had ERRORS

these appear to be unrelated to the PR changes

>

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17801/18502/summary.html

slava77 commented:
I notice that the edmConfigDump of step3 in 136.725 now takes a bit longer and it produces 70% more lines.

slava77 commented:
@davidlange6 
I expect that this and #17524 /  #17720 will require some time to get in sync for both to be in the release.
The #17524 /  #17720 is needed in context of 80X, which also has a target of 910pre1

slava77 commented:
If I understand correctly the edmConfigDump output increase with this PR,
the modules which were purged by convertToUnscheduled or similar in the baseline are no longer removed from the global config phase space.
Since we have a new fashion now to loop over everything in the process to apply edits and customizations, this feature will slow down the python parsing considerably, especially for simpler test configs that load a lot but then define paths/sequences/tasks for a tiny subset.

fwyzard commented:
Meaning people should finally clean up the configurations and remove the
unused modules from the standard sequences.

slava77 commented:
Just to summarize:
- there are a few comments on the code changes that should improve maintainability and can be addressed in a follow up PR
- this currently interferes with one somewhat large pending PR  #17720 (it's review is not complete yet, due to \~x2 growth in size in the last 24 hours)
- there is a notable increase in the quantity of objects left floating in the python space when the configs are finally parsed https://github.com/cms-sw/cmssw/pull/17801#issuecomment-287341343 and https://github.com/cms-sw/cmssw/pull/17801#issuecomment-287265057 . This can probably be improved in a follow up PR
- at run time I see that there are fewer modules constructed (which is a good addition in this PR, which included some cleanup of loaded and not running: in 136.725 step3 the total goes down from 2020 to 1973 of which only 1597 actually run in the 200 events tested)
- there are no differences in outputs in jenkins or in local tests

So, at the moment, from reco POV, only interference with #17720 is the current issue.

slava77 commented:
On 3/17/17 5:32 AM, Andrea Bocci wrote:
> Meaning people should finally clean up the configurations and remove the
> unused modules from the standard sequences.
> 

This PR alone didn't help in that direction (check JetCorrectors import
*, replacing named individual import)

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17801#issuecomment-287342235>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcblmoiEBwyAlowIs_GAH35UHhKVyGks5rmn1TgaJpZM4MUjDj>.
>

wddgit commented:
I looked through PR #17720. Most of the code I will need to modify to use Tasks is either in or called by code in customizeMiniAOD_MuEGFixMoriond2017.py (that file plus the other code it calls is getting close to a thousand lines of code). One issue I have is that I cannot see anything in the PR that calls the customizeAll function that drives all this. Most of it is also completely new code and new functions so there is also nothing in the release calling it. Is there a test already or can someone create a test that I could use to test this as I make my modifications? Seems like a test would be needed anyway? Or maybe I just missed it or one of the runTheMatrix tests will exercise this somehow.

slava77 commented:
On 3/17/17 1:05 PM, W. David Dagenhart wrote:
> I looked through PR #17720 <https://github.com/cms-sw/cmssw/pull/17720>.
> Most of the code I will need to modify to use Tasks is either in or
> called by code in customizeMiniAOD_MuEGFixMoriond2017.py (that file plus
> the other code it calls is getting close to a thousand lines of code).
> One issue I have is that I cannot see anything in the PR that calls the
> customizeAll function that drives all this. Most of it is also
> completely new code and new functions so there is also nothing in the
> release calling it. Is there a test already or can someone create a test
> that I could use to test this as I make my modifications? Seems like a
> test would be needed anyway? Or maybe I just missed it or one of the
> runTheMatrix tests will exercise this somehow.
> 

The customizeAll is not called in this PR, but it will need to be maintained

a sample cmsDriver command on top of e.g. 136.725 is

cmsDriver.py step4  --conditions auto:run2_data_relval -s
PAT,DQM:@miniAODDQM --runUnscheduled  --process PAT --data  --era
Run2_2016 --eventcontent MINIAOD,DQM --hltProcess reHLT --scenario pp
--datatier MINIAOD,DQM --customise_unsch
PhysicsTools/PatAlgos/slimming/customizeMiniAOD_MuEGFixMoriond2017.customizeAll
-n 200 --customise Validation/Performance/TimeMemoryInfo.py --filein
file:step3.root  --fileout file:step4.root --no_exec




> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17801#issuecomment-287457685>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbnkVBNNCZ_fbTDdaC8zHg3CfPRs0ks5rmud_gaJpZM4MUjDj>.
>

slava77 commented:
@wddgit 
there are merge conflicts in this PR now.

wddgit commented:
I will fix the conflict and rebase. I was already working on the conflicts and issues that will come when PR #17720 gets merged. Do you know when that will happen? If it is soon, I will wait for that and only rebase once.

slava77 commented:
On 3/20/17 9:03 AM, W. David Dagenhart wrote:
> I will fix the conflict and rebase. I was already working on the
> conflicts and issues that will come when PR #17720
> <https://github.com/cms-sw/cmssw/pull/17720> gets merged. Do you know
> when that will happen? If it is soon, I will wait for that and only
> rebase once.
> 

#17720 has unresolved issues from review and the main developer is not
available this week (Moriond, I guess).
If 910pre1 is ready to go for all other reasons, I think that this PR
(#17801) can go in first.


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/17801#issuecomment-287807222>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbrPjhvHKDrGYZfYk4lCTBTKeVP-Dks5rnqNfgaJpZM4MUjDj>.
>

wddgit committed:
Update reference file of test after rebase

cmsbuild commented:
Pull request #17801 was updated. @perrotta, @smuzaffar, @Dr15Jones, @silviodonato, @fwyzard, @franzoni, @dmitrijus, @Martin-Grunewald, @cmsbuild, @slava77, @vanbesien, @monttj, @davidlange6 can you please check and sign again.

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

wddgit commented:
I rebased to same release as IB CMSSW_9_1_X_2017-03-20-1100 (dcef19f7),
fixed a conflict and updated a reference file used by a unit test.

I noticed one new issue since the last rebase that I did not fix.
There is a new file:
RecoPixelVertexing/PixelLowPtUtilities/test/ClusterShapeExtractor_Phase2_cfg.py
The README.md file in the same directory says this is a reference
partial configuration that a user is supposed to edit manually before running.
Given the way this is setup, I think the author should migrate this one to use
Tasks. I mention it because after this PR is merged it will be broken.
It does not appear to be a unit test. It looks to be derived from the output
of cmsDriver, so they should be able to see how to fix it by looking
at how the cmsDriver output changes. Author was @kandrosov PR #17804

As far as I am concerned, this is ready to be merged if the PR testing
passes. I think all issues in the comments can be handled in subsequent
PRs.

cmsbuild commented:
-1

Tested at: dc6030ab877c731f4fce2a5a727ac9bb7af46ed6

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17801/18577/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testPythonParameterSet had ERRORS

cmsbuild commented:
Comparison job queued.

Dr15Jones commented:
The unit test failure is unrelated to this pull request.

Dr15Jones commented:
+1

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17801/18577/summary.html

davidlange6 commented:
Hi all
I'll merge this in ~3 hours.. please comment on any remaining concerns.

Dr15Jones commented:
The champagne bottle has been opened!

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/17801	dc6030a
- changes are in line with the description including edits in many reco-related files 
- jenkins tests pass and comparisons with baseline show no differences
- local tests with an earlier version show reasonable/expected technical behavior https://github.com/cms-sw/cmssw/pull/17801#issuecomment-287342963

The expectation  is that issues mentioned in review of this PR are resolved as a part of #18018
